src/Open3D/Visualization/Rendering/Filament/FilamentResourceManager.cpp

```diff
@@ -0,0 +1,718 @@
+        case 1:
+            settings.imageType = filament::Texture::Type::UBYTE;
+            break;
+        default:
+            utility::LogError("Unsupported image bytes per channel: {}",
+                              image.bytes_per_channel_);
+            break;
+    }
+
+    return settings;
+}
+}  // namespace texture_loading
+
+}  // namespace
+
+const MaterialHandle FilamentResourceManager::kDefaultLit =
+        MaterialHandle::Next();
+const MaterialHandle FilamentResourceManager::kDefaultUnlit =
+        MaterialHandle::Next();
+const MaterialInstanceHandle FilamentResourceManager::kDepthMaterial =
+        MaterialInstanceHandle::Next();
+const MaterialInstanceHandle FilamentResourceManager::kNormalsMaterial =
+        MaterialInstanceHandle::Next();
+const MaterialInstanceHandle FilamentResourceManager::kColorMapMaterial =
+        MaterialInstanceHandle::Next();
+const TextureHandle FilamentResourceManager::kDefaultTexture =
+        TextureHandle::Next();
+const TextureHandle FilamentResourceManager::kDefaultColorMap =
+        TextureHandle::Next();
+const TextureHandle FilamentResourceManager::kDefaultNormalMap =
+        TextureHandle::Next();
+
+static const std::unordered_set<REHandle_abstract> kDefaultResources = {
+        FilamentResourceManager::kDefaultLit,
+        FilamentResourceManager::kDefaultUnlit,
+        FilamentResourceManager::kDepthMaterial,
+        FilamentResourceManager::kNormalsMaterial,
+        FilamentResourceManager::kDefaultTexture,
+        FilamentResourceManager::kDefaultColorMap,
+        FilamentResourceManager::kDefaultNormalMap};
+
+FilamentResourceManager::FilamentResourceManager(filament::Engine& aEngine)
+    : engine_(aEngine) {
+    LoadDefaults();
+}
+
+FilamentResourceManager::~FilamentResourceManager() { DestroyAll(); }
+
+MaterialHandle FilamentResourceManager::CreateMaterial(const void* materialData,
+                                                       const size_t dataSize) {
+    using namespace filament;
+
+    Material* material =
+            Material::Builder().package(materialData, dataSize).build(engine_);
+
+    MaterialHandle handle;
+    if (material) {
+        handle =
+                RegisterResource<MaterialHandle>(engine_, material, materials_);
+    }
+
+    return handle;
+}
+
+MaterialHandle FilamentResourceManager::CreateMaterial(
+        const ResourceLoadRequest& request) {
+    MaterialHandle handle;
+
+    if (false == request.path.empty()) {
+        std::vector<char> materialData;
+        std::string errorStr;
+
+        if (utility::filesystem::FReadToBuffer(request.path, materialData,
+                                               &errorStr)) {
+            handle = CreateMaterial(materialData.data(), materialData.size());
+        } else {
+            request.errorCallback(request, errno, errorStr);
+        }
+    } else if (request.dataSize > 0) {
+        handle = CreateMaterial(request.data, request.dataSize);
```

Comment:
Filament throws an exception if it fails to create a material. I tried try/catch(...) but it didn't work, I assume because of issues catching exceptions across library boundaries. I don't think we can get a good solution for this before the release, so I added a TODO comment.

Reply:
this specifically has to do with Filament, it is using libc++ (everything else, us, python, pytorch, tensorflow, etc; uses libstdc++ on linux; on mac everything uses libc++ so there shouldn't be a problem); libc++ has its own symbols for exceptions, in addition it has different std::strings (incompatible with libstdc++ std::strings), so at least on linux there's a big mess around this, and if Filament throws exceptions then we will need to look into this; (all the other libraries, us, pytorch,tensorflow,python, we all throw and catch each others exceptions just fine)